src/controller_examples/rabbitmq_controller/proof/liveness/liveness.rs

```diff
@@ -68,52 +86,75 @@ proof fn next_with_wf_is_stable()
     RMQCluster::action_weak_fairness_is_stable(RMQCluster::disable_crash());
     RMQCluster::action_weak_fairness_is_stable(RMQCluster::disable_busy());
     stable_and_n!(
         always(lift_action(RMQCluster::next())),
         tla_forall(|input| RMQCluster::kubernetes_api_next().weak_fairness(input)),
         tla_forall(|input| RMQCluster::controller_next().weak_fairness(input)),
         tla_forall(|input| RMQCluster::schedule_controller_reconcile().weak_fairness(input)),
+        tla_forall(|input| RMQCluster::builtin_controllers_next().weak_fairness(input)),
         RMQCluster::disable_crash().weak_fairness(()),
         RMQCluster::disable_busy().weak_fairness(())
     );
 }
 
-// All assumptions that makes liveness possible, such as controller crash no longer happens,
-// the cr's spec always remains unchanged, and so on.
-spec fn assumptions(rabbitmq: RabbitmqClusterView) -> TempPred<RMQCluster> {
-    always(lift_state(RMQCluster::crash_disabled()))
-    .and(always(lift_state(RMQCluster::busy_disabled())))
+spec fn next_with_wf_and_assumption(rabbitmq: RabbitmqClusterView) -> TempPred<RMQCluster> {
+    always(lift_action(RMQCluster::next()))
+    .and(tla_forall(|input| RMQCluster::kubernetes_api_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::controller_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::schedule_controller_reconcile().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::builtin_controllers_next().weak_fairness(input)))
+    .and(RMQCluster::disable_crash().weak_fairness(()))
+    .and(RMQCluster::disable_busy().weak_fairness(()))
     .and(always(lift_state(RMQCluster::desired_state_is(rabbitmq))))
-    .and(always(lift_state(RMQCluster::the_object_in_schedule_has_spec_and_uid_as(rabbitmq))))
-    .and(always(lift_state(RMQCluster::the_object_in_reconcile_has_spec_and_uid_as(rabbitmq))))
 }
 
-proof fn assumptions_is_stable(rabbitmq: RabbitmqClusterView)
+spec fn next_with_wf_and_assumption_and_all_invariants(rabbitmq: RabbitmqClusterView) -> TempPred<RMQCluster> {
+    always(lift_action(RMQCluster::next()))
+    .and(tla_forall(|input| RMQCluster::kubernetes_api_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::controller_next().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::schedule_controller_reconcile().weak_fairness(input)))
+    .and(tla_forall(|input| RMQCluster::builtin_controllers_next().weak_fairness(input)))
+    .and(RMQCluster::disable_crash().weak_fairness(()))
+    .and(RMQCluster::disable_busy().weak_fairness(()))
+    .and(always(lift_state(RMQCluster::desired_state_is(rabbitmq))))
+    .and(invariants_phase_I(rabbitmq))
+    .and(invariants_phase_II(rabbitmq))
+    .and(invariants_phase_III(rabbitmq))
+    .and(invariants_phase_IV(rabbitmq))
+    .and(invariants_phase_V(rabbitmq))
+    .and(invariants_phase_VI(rabbitmq))
```

Comment:
I like the idea of explicitly splitting the proof into phases (we've been doing that but here you make it very explicit) and group invariants into phases. I think finding the right way to split the execution into phases that are amenable to proof is one of the keys to liveness verification.

Reply:
That's what I talk about in the description